### ISSUE TITLE
World Name API & Plugin World Storage

### ISSUE THREAD

**Issue by ryantheleach** [SpongePowered/SpongeAPI#1738]
@TheRaspPie https://github.com/SpongePowered/SpongeCommon/issues/1048
```
If you load a world with Server#loadWorld(String) or get its name with World#getName()
the expected/returned world name is the one from the internal NBT data, not the name of the world folder.
So if you rename the world folder like you were easily able to do with Bukkit, the world won't load with the new name. If you have tens of worlds which got renamed a long time ago and whose internal names differ from the folder names or even if you just want to rename the world today, you can't load it anymore.

Would it be possible to change the standard behaviour so that Sponge finds worlds by it's folder name or add a config option to change the behaviour?

Thanks in advance.
```

Current API
```
loadWorld -> Directory name
getName -> Directory name (??)
setName -> NBT name
UUID getUniqueId();
String getDefaultWorldName();
```

My Proposed API
```
loadWorld(WorldProperties) //Is this possible?
loadWorld(UUID) //bad. Causes ambiguity if people copy the directory.
loadWorld(Path) //Not terrible. Maybe too flexible
loadWorld(String directoryName) //No Ambiguity, but no prior parsing of if it will fail due to UUID clash
loadWorld(Some form of WorldProvider) ??? //Database provided worlds? FakeWorlds?
Text getDisplayName() //If this isn't a full Text in 1.13 rather then legacy string I'll eat my hat.
void setDisplayName(Text text) //might need conversion to legacy.
UUID getUniqueId();
String getLevelName(); // Named like this to disambiguate & imply it's what is set in server.properties. previously. getDefaultWorldName(); I think this is important in order to prevent people using it for creating a path.
Path getPluginWorldStorage(WorldProperties, Plugin) //Allows for alternate world implementations to still have a World directory that plugins can store data in, regardless of if the world exists on disk.
```

I'm wondering if we should be avoiding exposing the world directory name directly, but if you do, that poses some issues for loading worlds.

Also some serious thought should be given about when the UUID is changed, if at all.

E.g. it should be an explicit admin action after importing a duplicate world to permanently change the UUID.

Usecase:

1. You are loading a backup of a world to use as a creative / test world permanently. You want it to copy all previous plugin data, but be able to access both your old world and this one simultaneously.
2. You are loading a backup of a world to preview to make sure you didn't roll back too far, you want to just see the world temporarily before switching it for your main world, You may or may not want to rollback the player / plugin data associated with a world, depending on plugin / circumstance.
3. You run a server where the map resets periodically, you are loading the world to edit the template world for the next run to add custom content for your twitch streaming event.

Basically what I'm saying is, if you want plugin data to exist and be copyOnWorldCopy you need to either associate with the UUID and deal with Sponge not loading 2 worlds with a duplicate ID, or changing one of them without telling you, or having 2 duplicate ID worlds loaded, or hope you listen to a duplicateUUIDWorld event or something.

Or ignore all of that, and store the data in a folder inside the world or dimension directory, or on world properties as a dataview, creating the association with the world reference on load.

**Comment by Zidane** [comment]
Couple of things:

- WorldProperties#getName now is the directory name which fixes a lot of the inconsistency
- At this time I don't want to let the API control where World storage occurs especially since that will be HELL to get MCP working right with that
- I might change the display name to a Text object, let me think on if that'll cause any problems.

My latest 1.14 work covers most of the other stuff.
